winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -253,458 +135,416 @@ static krb5_error_code krb5_prompter(krb5_context context, void* data, const cha
-		goto cleanup;
+		/* Use the default cache with it's default principal */
+		if ((rv = krb5_cc_default(ctx, &ccache)))
+			goto cleanup;
+		if ((rv = krb5_cc_get_principal(ctx, ccache, &principal)))
+			goto cleanup;
 	}
-
-	/* Get credentials */
-	if ((ret = krb5_init_creds_get(ctx, init_ctx)))
+	else
 	{
-		krb5_log_error_message(ctx, ret, WLOG_ERROR, "error while getting credentials");
-		goto cleanup;
+		if ((rv = krb5_cc_resolve(ctx, "MEMORY:", &ccache)))
+			goto cleanup;
 	}
 
-	/* Retrieve credentials */
-	if ((ret = krb5_init_creds_get_creds(ctx, init_ctx, &creds)))
+	if (krb_settings && krb_settings->keytab)
 	{
-		krb5_log_error_message(ctx, ret, WLOG_ERROR, "error while retrieving credentials");
-		goto cleanup;
+		if ((rv = krb5_kt_resolve(ctx, krb_settings->keytab, &keytab)))
+			goto cleanup;
 	}
-
-#ifdef WITH_GSSAPI_HEIMDAL
-
-	/* For Heimdal, we use this function to store credentials */
-	if ((ret = krb5_init_creds_store(ctx, init_ctx, ccache)))
+	else
 	{
-		krb5_log_error_message(ctx, ret, WLOG_ERROR, "error while storing credentials");
+		if ((rv = krb5_kt_default(ctx, &keytab)))
+			goto cleanup;
+	}
+	
+	if ((rv = krb5_get_init_creds_opt_alloc(ctx, &gic_opt)))
 		goto cleanup;
+
+	krb5_get_init_creds_opt_set_forwardable(gic_opt, 0);
+	krb5_get_init_creds_opt_set_proxiable(gic_opt, 0);
+
+	if (krb_settings)
+	{
+		if (krb_settings->startTime)
+			start_time = krb_settings->startTime;
+		if (krb_settings->lifeTime)
+			krb5_get_init_creds_opt_set_tkt_life(gic_opt, krb_settings->lifeTime);
+		if (krb_settings->renewLifeTime)
+			krb5_get_init_creds_opt_set_renew_life(gic_opt, krb_settings->renewLifeTime);
+		if (krb_settings->withPac)
+			krb5_get_init_creds_opt_set_pac_request(ctx, gic_opt, TRUE);
+		if (krb_settings->armorCache)
+			krb5_get_init_creds_opt_set_fast_ccache_name(ctx, gic_opt, krb_settings->armorCache);
+		if (krb_settings->pkinitX509Identity)
+			krb5_get_init_creds_opt_set_pa(ctx, gic_opt, "X509_user_identity",
+			                               krb_settings->pkinitX509Identity);
+		if (krb_settings->pkinitX509Anchors)
+			krb5_get_init_creds_opt_set_pa(ctx, gic_opt, "X509_anchors",
+			                               krb_settings->pkinitX509Anchors);
 	}
 
+#ifdef WITH_GSSAPI_MIT
+	krb5_get_init_creds_opt_set_out_ccache(ctx, gic_opt, ccache);
+	krb5_get_init_creds_opt_set_in_ccache(ctx, gic_opt, ccache);
 #endif
-cleanup:
-	krb5_free_cred_contents(ctx, &creds);
-#ifdef HAVE_AT_LEAST_KRB_V1_13
 
-	/* MIT Kerberos version 1.13 at minimum.
-	 * For releases 1.12 and previous, krb5_get_init_creds_opt structure
-	 * is freed in krb5_init_creds_free() */
-	if (options)
-		krb5_get_init_creds_opt_free(ctx, options);
+	krb5_cc_get_full_name(ctx, ccache, &ccache_name);
+	krb5_kt_get_name(ctx, keytab, keytab_name, PATH_MAX);
 
-#endif
-
-	if (init_ctx)
-		krb5_init_creds_free(ctx, init_ctx);
+	cred_store_opts[0].key = "ccache";
```

Comment:
I'm not sure that it's a so good idea to provide the keytab option because if `keytab_name` is set but the file not accessible for some reason, it stops the call to `gss_acquire_cred_from()` (I've looked at the source code).
The options given in `cred_store_opts` are a logical AND not a OR (all given options must be retrieved, unless the value is NULL).

Reply:
I suppose we can just provide it only if acceptor credentials are required

Reply:
Scratch that, it's only provided server side. But isn't it necessary to have the keytab anyway?

Reply:
Well the client could wish to log in with a keytab, even if it doesn't fit very well with RDP. 
I was wondering if the key name is `client_keytab` or `keytab` the doc is not very clear.

Reply:
I'm pretty sure it's `client_keytab`. We could add a keytab option client side, but then we'd run into the issue you raised